ykrt/src/compile/j2/opt/known_bits.rs

```diff
@@ -0,0 +1,482 @@
+//! Known bits analysis.
+//!
+//! Data-flow analysis to gain information about bits.
+//! We use similar ideas to PyPy:
+//! https://pypy.org/posts/2024/08/toy-knownbits.html
+//!
+
+use crate::compile::{
+    j2::{
+        hir::{And, Const, ConstKind, Inst, InstIdx, InstT, Or, Ty},
+        opt::{
+            BlockLikeT,
+            fullopt::{CommitInstOpt, OptOutcome, PassOpt, PassT},
+        },
+    },
+    jitc_yk::arbbitint::ArbBitInt,
+};
+use index_vec::IndexVec;
+
+/// Known-bits analysis.
+pub(super) struct KnownBits {
+    known_bits: IndexVec<InstIdx, Option<KnownBitValue>>,
+    pending_commit: Option<KnownBitValue>,
+}
+
+impl PassT for KnownBits {
+    fn feed(&mut self, opt: &mut PassOpt, inst: Inst) -> OptOutcome {
+        self.pending_commit = None;
+        match inst {
+            Inst::And(x) => self.opt_and(opt, x),
+            Inst::Const(x) => self.opt_const(x),
+            Inst::Or(x) => self.opt_or(opt, x),
+            _ => OptOutcome::Rewritten(inst),
+        }
+    }
+
+    fn inst_committed(&mut self, _opt: &CommitInstOpt, _iidx: InstIdx, _inst: &Inst) {
+        self.known_bits.push(self.pending_commit.clone());
+        assert_eq!(_iidx.index(), self.known_bits.len() - 1);
```

Comment:
`iidx` doesn't need the leading `_` if it's used in `assert`. Also, wouldn't this `assert` be clearer _before_ the push when the `- 1` wouldn't be needed?